services/user/Subgroups/service/src/lib.rs

```diff
@@ -0,0 +1,104 @@
+#[psibase::service(name = "subgroups")]
+pub mod service {
+    pub use psibase::fracpack;
+
+    /// Function that determines which subgrouping solution is preferred based on specific criteria.
+    /// Returns true if the new solution is preferred over the old solution.
+    type PreferenceFunction = fn(new: &[u32], old: &[u32]) -> bool;
+
+    /// Preference function that compares based on the greatest minimum group size.
+    /// If minimum values are equal, it compares based on maximum group size.
+    fn greatest_minimum_preference(new: &[u32], old: &[u32]) -> bool {
+        let old_min = old.iter().min().unwrap_or(&u32::MAX);
+        let new_min = new.iter().min().unwrap_or(&u32::MAX);
+
+        // Criterion 1: If the new minimum is greater than the old minimum
+        let criterion1 = new_min > old_min;
+
+        // Criterion 2: If the minimum values are equal, then prefer the solution
+        // with the minimum number of groups at the minimum group size
+        let criterion2 = new_min == old_min
+            && new.iter().filter(|&x| x == new_min).count()
+                < old.iter().filter(|&x| x == old_min).count();
+
+        criterion1 || criterion2
+    }
+
+    /// Finds a valid partition of `n` into groups of `allowed` sizes
+    ///
+    /// Parameters
+    /// * `n` - The number to partition (population)
+    /// * `allowed` - Vector of allowed group sizes
+    /// * `is_preferred` - Function that determines which solution is preferred
```

Comment:
The requirements for `is_preferred` are
- It must define a strict weak ordering for partitions of each total size, up to `n`.
- Adding the same group to two partitions must not change the result of `is_preferred`

Reply:
Added these to the doc comment